test/libsolidity/analysis/FunctionCallGraph.cpp

```diff
@@ -173,8 +173,8 @@ void checkCallGraphExpectations(
 
 ostream& operator<<(ostream& _out, EdgeNames const& _edgeNames)
 {
-	for (auto const& edge: _edgeNames | to<vector>() | actions::sort(std::less()))
-		_out << "    " << get<0>(edge) << " -> " << get<1>(edge) << endl;
+	for (auto&& [from, to]: _edgeNames)
```

Comment:
Why not `const &`? Is this actually `&&` reference here?

Reply:
That's a use of `&&` that some people call "universal reference"... it basically just tells the compiler, I don't need a copy, do whatever is best. In particular since `edgeNames` is a const ref, the tuple elements will be const refs, so this will reduce to rref to const ref which is just const ref...
So this is actually just being lazy to think about what to put there, because `&&` pretty much always works and always does what you want, hence using it for this is quite idiomatic I'd say.

But, that being said, since we now did have a look at it, I might as well change it to an explicit ``const&``, even though it's the same ;-).

Reply:
Thanks for the explanation :)

For me the advantage of `const&` is that it's read-only so you get an error if you modify it by accident. I see that in case of a `tuple` like here I can't modify it anyway but I could if it was for example an `int`. So I'd prefer `const&&`.

Also, I haven't had that many opportunities to use `&&` so it makes me stop and think about the difference. But, well, I should probably get used to it :)

Reply:
I'd actually be careful with stuff like ``const&&`` :-). It's probably best to either just use plain `&&` (basically saying: I don't care, dear compiler, do what you think is best) or making the actual reference kind explicit.
The reason being that once you do anything else you actually *have* to think about what this stuff will reduce to...
``auto`` might be deduced as ``T const&``, then you have to match this with ``T const& const&&``... maybe that's fine and gives you what you want... but what happens in other cases where ``auto`` is actually reduced to ``T&``? That would yield ``T& const&&``... Will that collapse to ``T&`` or ``T const&``? I'd actually need to check :-D.

Reply:
Right, good point. I suspect it would just be `const&` but yeah I also would prefer not to have to guess and risk that it's not.

In that case my vote is for `const&` unless the circumstances force me to use something else :)

Reply:
Yeah, actually ``auto const&&`` doesn't even match ``const&`` :-).

Reply:
Scott Meyers has mainly been pushing for this "universal reference" concept, I think, in case you're interested https://isocpp.org/blog/2012/11/universal-references-in-c11-scott-meyers

Reply:
You really don't need to guess, though - it really always does what you want it to do :-).